File: frontend/src/lib/networth.test.ts
import { describe, it, expect } from "vitest";
import {
    latestBalanceHome,
    summarizeAccounts,
    cashChartAccountsOf,
    sampleProjection,
    netWorthBreakdown,
    type AccountLike,
} from "./networth";
import { AccountKind, LiquidityStatus } from "../types/types";
import type { BalanceResponse, NetWorthProjectionPoint } from "../types/types";

// The scenario these guard: a household with a 500k home, a 400k mortgage and
// 10k in the bank. Net worth is +110k. Before property existed the same
// household read -390k, which is the bug this feature exists to fix.

const bal = (date: string, amount: number, home?: number): BalanceResponse => ({
    id: `${date}-${amount}`,
    account_id: "a",
    date,
    balance: amount,
    balance_home_currency: home ?? amount,
    is_manual: true,
});

const account = (over: Partial<AccountLike>): AccountLike => ({
    id: Math.random().toString(),
    kind: AccountKind.Asset,
    liquidity: LiquidityStatus.Liquid,
    currency: "USD",
    history: [],
    ...over,
});

describe("latestBalanceHome", () => {
    it("picks the newest checkpoint regardless of array order", () => {
        expect(latestBalanceHome([bal("2026-01-01", 100), bal("2026-06-01", 900), bal("2026-03-01", 500)]))
            .toBe(900);
    });

    it("prefers the home-currency value over the native one", () => {
        expect(latestBalanceHome([bal("2026-01-01", 1000, 750)])).toBe(750);
    });

    it("returns 0 for an empty history instead of NaN", () => {
        expect(latestBalanceHome([])).toBe(0);
    });

    it("coerces a non-numeric balance to 0 rather than poisoning aggregates", () => {
        const broken = { ...bal("2026-01-01", 0), balance: undefined as unknown as number, balance_home_currency: undefined };
        expect(latestBalanceHome([broken])).toBe(0);
    });
});

describe("summarizeAccounts", () => {
    const home = account({
        liquidity: LiquidityStatus.Illiquid,
        history: [bal("2026-01-01", 500_000)],
    });
    const mortgage = account({
        kind: AccountKind.Liability,
        history: [bal("2026-01-01", 400_000)],
    });
    const savings = account({ history: [bal("2026-01-01", 10_000)] });

    it("counts property in net worth alongside the loan against it", () => {
        const totals = summarizeAccounts([home, mortgage, savings]);
        expect(totals.totalAssets).toBe(510_000);
        expect(totals.liabilities).toBe(400_000);
        expect(totals.net).toBe(110_000);
    });

    it("keeps property out of liquid-now", () => {
        const totals = summarizeAccounts([home, savings]);
        expect(totals.property).toBe(500_000);
        expect(totals.liquidNow).toBe(10_000);
    });

    it("subtracts a liability rather than adding it", () => {
        expect(summarizeAccounts([mortgage]).net).toBe(-400_000);
    });

    it("groups time-locked and retirement accounts together", () => {
        const cpf = account({ liquidity: LiquidityStatus.TimeLocked, history: [bal("2026-01-01", 30_000)] });
        const srs = account({ liquidity: LiquidityStatus.Retirement, history: [bal("2026-01-01", 20_000)] });
        expect(summarizeAccounts([cpf, srs]).retirement).toBe(50_000);
    });

    it("reports distinct currencies, sorted", () => {
        const sgd = account({ currency: "SGD" });
        const usd = account({ currency: "USD" });
        expect(summarizeAccounts([usd, sgd, usd]).currencies).toEqual(["SGD", "USD"]);
    });

    it("returns zeroes for an empty household", () => {
        const totals = summarizeAccounts([]);
        expect(totals).toMatchObject({ totalAssets: 0, liabilities: 0, net: 0, property: 0 });
    });
});

describe("summarizeAccounts with money owed", () => {
    // The hole this fills: splitting a $120 bill takes $120 out of the bank and
    // records $40 of spending. Without the receivable, net worth reports the
    // other $80 as having simply evaporated.
    const savings = account({ history: [bal("2026-01-01", 1_000)] });

    it("counts money owed to you as an asset", () => {
        const totals = summarizeAccounts([savings], { owedToYou: 80, youOwe: 0 });
        expect(totals.totalAssets).toBe(1_080);
        expect(totals.net).toBe(1_080);
        expect(totals.receivables).toBe(80);
    });

    it("keeps money owed to you out of liquid now", () => {
        // You cannot spend it this week. Treating it as spendable is how a
        // runway starts lying — the same reason property is excluded.
        const totals = summarizeAccounts([savings], { owedToYou: 80, youOwe: 0 });
        expect(totals.liquidNow).toBe(1_000);
    });

    it("counts money you owe as a liability", () => {
        const totals = summarizeAccounts([savings], { owedToYou: 0, youOwe: 45 });
        expect(totals.liabilities).toBe(45);
        expect(totals.net).toBe(955);
    });

    it("nets the two directions into net worth without merging them", () => {
        const totals = summarizeAccounts([savings], { owedToYou: 80, youOwe: 45 });
        expect(totals.receivables).toBe(80);
        expect(totals.liabilities).toBe(45);
        expect(totals.net).toBe(1_035);
    });

    it("changes nothing when nobody owes anybody", () => {
        expect(summarizeAccounts([savings])).toEqual(summarizeAccounts([savings], { owedToYou: 0, youOwe: 0 }));
    });

    it("survives a non-finite total rather than poisoning net worth", () => {
        const totals = summarizeAccounts([savings], { owedToYou: NaN, youOwe: 45 });
        expect(totals.totalAssets).toBe(1_000);
        expect(totals.net).toBe(955);
    });
});

describe("netWorthBreakdown with money owed", () => {
    const savings = account({ history: [bal("2026-01-01", 1_000)] });

    it("gives money owed its own slice rather than burying it in Other", () => {
        const breakdown = netWorthBreakdown([savings], 0, { owedToYou: 80, youOwe: 0 });
        const owed = breakdown.slices.find(s => s.key === "owed");
        expect(owed?.value).toBe(80);
        expect(breakdown.slices.find(s => s.key === "other")).toBeUndefined();
    });

    it("omits the slice when nobody owes you", () => {
        const breakdown = netWorthBreakdown([savings], 0);
        expect(breakdown.slices.find(s => s.key === "owed")).toBeUndefined();
    });
});

describe("cashChartAccountsOf", () => {
    it("excludes property and liabilities, keeping spendable assets", () => {
        const accounts = [
            { id: "cash", kind: AccountKind.Asset, liquidity: LiquidityStatus.Liquid },
            { id: "stocks", kind: AccountKind.Asset, liquidity: LiquidityStatus.MarketLiquid },
            { id: "house", kind: AccountKind.Asset, liquidity: LiquidityStatus.Illiquid },
            { id: "loan", kind: AccountKind.Liability, liquidity: LiquidityStatus.Liquid },
        ];
        expect(cashChartAccountsOf(accounts).map(a => a.id)).toEqual(["cash", "stocks"]);
    });
});

describe("netWorthBreakdown", () => {
    it("buckets cash, retirement and property, and adds investments as their own slice", () => {
        const savings = account({ history: [bal("2026-01-01", 10_000)] });
        const cpf = account({ liquidity: LiquidityStatus.TimeLocked, history: [bal("2026-01-01", 30_000)] });
        const home = account({ liquidity: LiquidityStatus.Illiquid, history: [bal("2026-01-01", 500_000)] });

        const { slices, sliceTotal } = netWorthBreakdown([savings, cpf, home], 20_000);

        expect(slices).toEqual([
            { key: "cash", label: "Cash", value: 10_000 },
            { key: "investments", label: "Investments", value: 20_000 },
            { key: "retirement", label: "Retirement & locked", value: 30_000 },
            { key: "property", label: "Property", value: 500_000 },
        ]);
        expect(sliceTotal).toBe(560_000);
    });

    it("reports liabilities alongside the slices rather than as a negative wedge", () => {
        const mortgage = account({ kind: AccountKind.Liability, history: [bal("2026-01-01", 400_000)] });
        const home = account({ liquidity: LiquidityStatus.Illiquid, history: [bal("2026-01-01", 500_000)] });

        const { slices, liabilities } = netWorthBreakdown([mortgage, home], 0);

        expect(liabilities).toBe(400_000);
        expect(slices.find(s => s.key === "liabilities")).toBeUndefined();
    });

    it("drops a negative bucket (an overdrawn household) instead of an impossible negative wedge", () => {
        const overdrawn = account({ history: [bal("2026-01-01", -12_000)] });
        const home = account({ liquidity: LiquidityStatus.Illiquid, history: [bal("2026-01-01", 500_000)] });

        const { slices, sliceTotal } = netWorthBreakdown([overdrawn, home], 0);

        expect(slices).toEqual([{ key: "property", label: "Property", value: 500_000 }]);
        // sliceTotal is the denominator for each slice's %, so it must match
        // the visible slices — not gross assets, which the excluded negative
        // cash bucket would otherwise drag below the property total.
        expect(sliceTotal).toBe(500_000);
    });

    it("buckets a market_liquid account as 'other' rather than dropping it", () => {
        const brokerageCash = account({ liquidity: LiquidityStatus.MarketLiquid, history: [bal("2026-01-01", 5_000)] });
        const { slices } = netWorthBreakdown([brokerageCash], 0);
        expect(slices).toEqual([{ key: "other", label: "Other assets", value: 5_000 }]);
    });
});

describe("sampleProjection", () => {
    const point = (date: string, net: number, liabilities = 0): NetWorthProjectionPoint => ({
        date,
        assets: net + liabilities,
        liabilities,
        net_worth: net,
    });

    // 25 monthly points: two full years plus one.
    const monthly = (count: number, fn: (i: number) => NetWorthProjectionPoint) =>
        Array.from({ length: count }, (_, i) => fn(i));

    it("thins monthly points down to yearly and keeps the final one", () => {
        const points = monthly(25, i => point(`2026-${String((i % 12) + 1).padStart(2, "0")}-01`, i * 1000, 400_000 - i * 1000));
        const sampled = sampleProjection(points);
        // indices 0, 12, 24 — and 24 is also the last point, not duplicated.
        expect(sampled).toHaveLength(3);
        expect(sampled[sampled.length - 1].netWorth).toBe(24_000);
    });

    it("negates liabilities so debt renders below the zero line", () => {
        const points = monthly(13, i => point("2026-01-01", -100, 400_000 - i));
        expect(sampleProjection(points)[0].liabilities).toBe(-400_000);
    });

    it("says nothing when there is no debt and nothing changes", () => {
        const points = monthly(13, () => point("2026-01-01", 25_000));
        expect(sampleProjection(points)).toEqual([]);
    });

    it("still reports a debt-free household whose assets are growing", () => {
        const points = monthly(13, i => point("2026-01-01", 25_000 + i * 100));
        expect(sampleProjection(points).length).toBeGreaterThan(0);
    });

    it("handles an empty or single-point projection", () => {
        expect(sampleProjection([])).toEqual([]);
        expect(sampleProjection([point("2026-01-01", 100)])).toEqual([]);
    });
});

File: android/app/src/test/java/com/ivanlee/financetracker/NetWorthTest.kt
package com.ivanlee.financetracker

import com.ivanlee.financetracker.logic.selectableAccounts
import com.ivanlee.financetracker.data.model.AccountResponse
import com.ivanlee.financetracker.data.model.BalanceResponse
import com.ivanlee.financetracker.data.model.LiquidityStatus
import com.ivanlee.financetracker.logic.NetWorthAccountInput
import com.ivanlee.financetracker.logic.latestBalanceHome
import com.ivanlee.financetracker.logic.OwedTotals
import com.ivanlee.financetracker.logic.netWorthBreakdown
import com.ivanlee.financetracker.logic.summarizeAccounts
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Assert.assertFalse
import org.junit.Test
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneOffset

/**
 * Kotlin twin of iOS's `NetWorthTests` / web's `networth.test.ts`.
 *
 * Backs the Dashboard's net worth total and its Net Worth Split donut, so a mistake here
 * misstates money on the first screen a user sees.
 */
class NetWorthTest {

    private fun day(iso: String): Instant = LocalDate.parse(iso).atStartOfDay(ZoneOffset.UTC).toInstant()

    private fun balance(amount: Double, home: Double? = null, date: String = "2026-01-01") =
        BalanceResponse(
            id = "$date-$amount",
            accountId = "a",
            date = day(date),
            balance = amount,
            balanceHomeCurrency = home ?: amount,
            isManual = true,
        )

    private fun account(
        kind: String? = "asset",
        liquidity: LiquidityStatus = LiquidityStatus.LIQUID,
        history: List<BalanceResponse> = emptyList(),
    ) = NetWorthAccountInput(kind, liquidity, history)

    // ---- latestBalanceHome ---------------------------------------------------------------

    @Test
    fun `latestBalanceHome picks the newest checkpoint`() {
        val value = latestBalanceHome(
            listOf(
                balance(100.0, date = "2026-01-01"),
                balance(900.0, date = "2026-06-01"),
                balance(500.0, date = "2026-03-01"),
            )
        )
        assertEquals(900.0, value, 0.0)
    }

    @Test
    fun `latestBalanceHome returns zero for empty history`() {
        assertEquals(0.0, latestBalanceHome(emptyList()), 0.0)
    }

    // ---- summarizeAccounts ----------------------------------------------------------------

    @Test
    fun `summarizeAccounts counts property alongside the loan against it`() {
        val home = account(liquidity = LiquidityStatus.ILLIQUID, history = listOf(balance(500_000.0)))
        val mortgage = account(kind = "liability", history = listOf(balance(400_000.0)))
        val savings = account(history = listOf(balance(10_000.0)))

        val totals = summarizeAccounts(listOf(home, mortgage, savings))
        assertEquals(510_000.0, totals.totalAssets, 0.0)
        assertEquals(400_000.0, totals.liabilities, 0.0)
        assertEquals(110_000.0, totals.net, 0.0)
    }

    @Test
    fun `summarizeAccounts keeps property out of liquid now`() {
        val home = account(liquidity = LiquidityStatus.ILLIQUID, history = listOf(balance(500_000.0)))
        val savings = account(history = listOf(balance(10_000.0)))

        val totals = summarizeAccounts(listOf(home, savings))
        assertEquals(500_000.0, totals.property, 0.0)
        assertEquals(10_000.0, totals.liquidNow, 0.0)
    }

    @Test
    fun `summarizeAccounts groups time locked and retirement together`() {
        val cpf = account(liquidity = LiquidityStatus.TIME_LOCKED, history = listOf(balance(30_000.0)))
        val srs = account(liquidity = LiquidityStatus.RETIREMENT, history = listOf(balance(20_000.0)))
        assertEquals(50_000.0, summarizeAccounts(listOf(cpf, srs)).retirement, 0.0)
    }

    // ---- money owed ---------------------------------------------------------------------------
    //
    // The hole this fills: splitting a $120 bill takes $120 out of the bank and records $40 of
    // spending. Without the receivable, net worth reports the other $80 as having evaporated.

    @Test
    fun `money owed to you counts as an asset`() {
        val savings = account(history = listOf(balance(1_000.0)))
        val totals = summarizeAccounts(listOf(savings), OwedTotals(owedToYou = 80.0))
        assertEquals(1_080.0, totals.totalAssets, 0.0)
        assertEquals(1_080.0, totals.net, 0.0)
        assertEquals(80.0, totals.receivables, 0.0)
    }

    @Test
    fun `money owed to you stays out of liquid now`() {
        // You cannot spend it this week. Treating it as spendable is how a runway starts
        // lying — the same reason property is excluded.
        val savings = account(history = listOf(balance(1_000.0)))
        val totals = summarizeAccounts(listOf(savings), OwedTotals(owedToYou = 80.0))
        assertEquals(1_000.0, totals.liquidNow, 0.0)
    }

    @Test
    fun `money you owe counts as a liability`() {
        val savings = account(history = listOf(balance(1_000.0)))
        val totals = summarizeAccounts(listOf(savings), OwedTotals(youOwe = 45.0))
        assertEquals(45.0, totals.liabilities, 0.0)
        assertEquals(955.0, totals.net, 0.0)
    }

    @Test
    fun `the two directions net into net worth without merging`() {
        val savings = account(history = listOf(balance(1_000.0)))
        val totals = summarizeAccounts(listOf(savings), OwedTotals(owedToYou = 80.0, youOwe = 45.0))
        assertEquals(80.0, totals.receivables, 0.0)
        assertEquals(45.0, totals.liabilities, 0.0)
        assertEquals(1_035.0, totals.net, 0.0)
    }

    @Test
    fun `nothing changes when nobody owes anybody`() {
        val savings = account(history = listOf(balance(1_000.0)))
        assertEquals(
            summarizeAccounts(listOf(savings)).net,
            summarizeAccounts(listOf(savings), OwedTotals.NONE).net,
            0.0,
        )
    }

    @Test
    fun `a non-finite total does not poison net worth`() {
        val savings = account(history = listOf(balance(1_000.0)))
        val totals = summarizeAccounts(
            listOf(savings),
            OwedTotals(owedToYou = Double.NaN, youOwe = 45.0),
        )
        assertEquals(1_000.0, totals.totalAssets, 0.0)
        assertEquals(955.0, totals.net, 0.0)
    }

    @Test
    fun `money owed gets its own slice rather than burying it in other`() {
        val savings = account(history = listOf(balance(1_000.0)))
        val breakdown = netWorthBreakdown(
            listOf(savings),
            portfolioValue = 0.0,
            owed = OwedTotals(owedToYou = 80.0),
        )
        assertEquals(80.0, breakdown.slices.first { it.key == "owed" }.value, 0.0)
        assertTrue(breakdown.slices.none { it.key == "other" })
    }

    @Test
    fun `the owed slice is omitted when nobody owes you`() {
        val savings = account(history = listOf(balance(1_000.0)))
        val breakdown = netWorthBreakdown(listOf(savings), portfolioValue = 0.0)
        assertTrue(breakdown.slices.none { it.key == "owed" })
    }

    // ---- netWorthBreakdown ------------------------------------------------------------------

    @Test
    fun `buckets cash retirement and property and adds investments as their own slice`() {
        val savings = account(history = listOf(balance(10_000.0)))
        val cpf = account(liquidity = LiquidityStatus.TIME_LOCKED, history = listOf(balance(30_000.0)))
        val home = account(liquidity = LiquidityStatus.ILLIQUID, history = listOf(balance(500_000.0)))

        val breakdown = netWorthBreakdown(listOf(savings, cpf, home), portfolioValue = 20_000.0)

        assertEquals(listOf("cash", "investments", "retirement", "property"), breakdown.slices.map { it.key })
        assertEquals(listOf(10_000.0, 20_000.0, 30_000.0, 500_000.0), breakdown.slices.map { it.value })
        assertEquals(560_000.0, breakdown.sliceTotal, 0.0)
    }

    @Test
    fun `reports liabilities alongside the slices rather than as a negative wedge`() {
        val mortgage = account(kind = "liability", history = listOf(balance(400_000.0)))
        val home = account(liquidity = LiquidityStatus.ILLIQUID, history = listOf(balance(500_000.0)))

        val breakdown = netWorthBreakdown(listOf(mortgage, home), portfolioValue = 0.0)

        assertEquals(400_000.0, breakdown.liabilities, 0.0)
        assertFalse(breakdown.slices.any { it.key == "liabilities" })
    }

    @Test
    fun `drops a negative bucket instead of an impossible negative wedge`() {
        val overdrawn = account(history = listOf(balance(-12_000.0)))
        val home = account(liquidity = LiquidityStatus.ILLIQUID, history = listOf(balance(500_000.0)))

        val breakdown = netWorthBreakdown(listOf(overdrawn, home), portfolioValue = 0.0)

        assertEquals(listOf("property"), breakdown.slices.map { it.key })
        // sliceTotal is the denominator for each slice's %, so it must match the visible
        // slices — not gross assets, which the excluded negative cash bucket would
        // otherwise drag below the property total.
        assertEquals(500_000.0, breakdown.sliceTotal, 0.0)
    }

    @Test
    fun `buckets a market liquid account as other rather than dropping it`() {
        val brokerageCash = account(liquidity = LiquidityStatus.MARKET_LIQUID, history = listOf(balance(5_000.0)))
        val breakdown = netWorthBreakdown(listOf(brokerageCash), portfolioValue = 0.0)
        assertEquals(listOf("other"), breakdown.slices.map { it.key })
        assertEquals(listOf(5_000.0), breakdown.slices.map { it.value })
    }
}

/**
 * Archived accounts — twin of the web `selectableAccounts` tests in
 * `lib/networth.test.ts` and iOS's `SelectableAccountsTests`.
 */
class SelectableAccountsTest {
    private fun account(id: String, archived: Boolean?) = AccountResponse(
        id = id,
        householdId = "h",
        name = "A",
        liquidity = LiquidityStatus.LIQUID,
        taxStatus = "taxable",
        kind = "asset",
        currency = "USD",
        isArchived = archived,
    )

    @Test
    fun `keeps archived accounts out of pickers`() {
        // Offering a closed account invites new activity on one the user has
        // finished with.
        val kept = selectableAccounts(listOf(account("a", false), account("b", true)))
        assertEquals(listOf("a"), kept.map { it.id })
    }

    @Test
    fun `treats a missing flag as open`() {
        // The field is optional on the wire; absent must never hide an account.
        assertEquals(1, selectableAccounts(listOf(account("c", null))).size)
    }
}
